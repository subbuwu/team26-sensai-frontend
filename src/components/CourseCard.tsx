import Link from "next/link";
import { useParams } from "next/navigation";
import { Trash2 } from "lucide-react";
import { useState } from "react";
import ConfirmationDialog from "./ConfirmationDialog";

interface CourseCardProps {
    course: {
        id: string | number;
        title: string;
        role?: string;
        org_id?: number;
        cohort_id?: number;
        org?: {
            slug: string;
        };
    };
    onDelete?: (courseId: string | number) => void;
}

export default function CourseCard({ course, onDelete }: CourseCardProps) {
    const params = useParams();
    const schoolId = params?.id;
    const [isDeleteConfirmOpen, setIsDeleteConfirmOpen] = useState(false);
    const [isDeleting, setIsDeleting] = useState(false);
    const [deleteError, setDeleteError] = useState<string | null>(null);

    // Generate a unique border color based on the course id
    const getBorderColor = () => {
        const colors = [
            'border-purple-500',
            'border-green-500',
            'border-pink-500',
            'border-yellow-500',
            'border-blue-500',
            'border-red-500',
            'border-indigo-500',
            'border-orange-500'
        ];

        // Handle string IDs by converting to a number
        let idNumber: number;
        if (typeof course.id === 'string') {
            // Use string hash code
            idNumber = Array.from(course.id).reduce(
                (hash, char) => ((hash << 5) - hash) + char.charCodeAt(0), 0
            );
            // Ensure positive number
            idNumber = Math.abs(idNumber);
        } else {
            idNumber = course.id;
        }

        return colors[idNumber % colors.length];
    };

    // Determine the correct link path
    const getLinkPath = () => {
        // If this is being viewed by a learner, use the school slug path
        if (course.role && course.role !== 'admin' && course.org?.slug) {
            // Include course_id and cohort_id as query parameters to help with selection on the school page
            return `/school/${course.org.slug}?course_id=${course.id}&cohort_id=${course.cohort_id}`;
        }
        // If we have an org_id from the API, use that for the school-specific course path
        else if (course.org_id) {
            return `/school/admin/${course.org_id}/courses/${course.id}`;
        }
        // If we're in a school context, use the school-specific course path
        else if (schoolId) {
            return `/school/admin/${schoolId}/courses/${course.id}`;
        }
        // Otherwise use the general course path
        return `/courses/${course.id}`;
    };

    // Check if this is an admin view
    const isAdminView = schoolId;

    const handleDeleteClick = (e: React.MouseEvent) => {
        e.preventDefault();
        e.stopPropagation();
        setIsDeleteConfirmOpen(true);
    };

    const handleDeleteConfirm = async () => {
        setIsDeleting(true);
        setDeleteError(null);

        try {
            const response = await fetch(`${process.env.NEXT_PUBLIC_BACKEND_URL}/courses/${course.id}`, {
                method: 'DELETE',
                headers: {
                    'Content-Type': 'application/json',
                }
            });

            if (!response.ok) {
                throw new Error('Failed to delete course');
            }

            // Close the dialog after successful deletion
            setIsDeleteConfirmOpen(false);

            // Call the onDelete callback if provided
            if (onDelete) {
                onDelete(course.id);
            }

        } catch (error) {
            console.error('Error deleting course:', error);
            setDeleteError('An error occurred while deleting the course. Please try again.');
        } finally {
            setIsDeleting(false);
        }
    };

    return (
        <div className="group relative">
            <Link href={getLinkPath()} className="block h-full">
                <div className={`bg-[#1A1A1A] text-gray-300 rounded-lg p-6 h-full transition-all hover:opacity-90 cursor-pointer border-b-2 ${getBorderColor()} border-opacity-70`}>
                    <h2 className="text-xl font-light mb-2">{course.title}</h2>
                </div>
            </Link>
            {isAdminView && (
                <button
                    className="absolute top-3 right-3 p-2 text-gray-400 hover:text-red-500 opacity-0 group-hover:opacity-100 transition-opacity focus:outline-none cursor-pointer rounded-full hover:bg-gray-800"
                    aria-label="Delete course"
                    onClick={handleDeleteClick}
                >
                    <Trash2 size={18} />
                </button>
            )}

            {/* Confirmation Dialog */}
            <ConfirmationDialog
                show={isDeleteConfirmOpen}
                title="Delete course"
                message={`Are you sure you want to delete this course? All the modules and tasks will be permanently deleted, any learner with access will lose all their progress and this action is irreversible`}
                confirmButtonText="Delete"
                onConfirm={handleDeleteConfirm}
                onCancel={() => setIsDeleteConfirmOpen(false)}
                type="delete"
                isLoading={isDeleting}
                errorMessage={deleteError}
            />
        </div>
    );
} 