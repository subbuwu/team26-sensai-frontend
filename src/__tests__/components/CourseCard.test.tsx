import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import CourseCard from '../../components/CourseCard';
import React from 'react';

// Mock next/navigation
jest.mock('next/navigation', () => ({
    useParams: jest.fn(),
}));

// Mock next/link
jest.mock('next/link', () => {
    return function MockLink({ children, ...props }: any) {
        return (
            <a {...props}>
                {children}
            </a>
        );
    };
});

// Mock fetch
const mockFetch = jest.fn();
global.fetch = mockFetch;

describe('CourseCard Component', () => {
    // Test data
    const basicCourse = {
        id: 123,
        title: 'Test Course'
    };

    const courseWithOrgId = {
        id: 456,
        title: 'Test Course with Org ID',
        org_id: 789
    };

    const courseWithRole = {
        id: 789,
        title: 'Learner Course',
        role: 'learner',
        org: {
            slug: 'test-org'
        },
        cohort_id: 123
    };

    beforeEach(() => {
        // Reset mocks before each test
        jest.clearAllMocks();

        // Default mock implementation for useParams
        require('next/navigation').useParams.mockReturnValue({ id: 'school-123' });

        // Mock successful fetch response
        mockFetch.mockResolvedValue({
            ok: true,
            json: async () => ({})
        });

        // Store original environment variables
        process.env.NEXT_PUBLIC_BACKEND_URL = 'https://api.example.com';
    });

    it('should render course title correctly', () => {
        render(<CourseCard course={basicCourse} />);
        expect(screen.getByText('Test Course')).toBeInTheDocument();
    });

    it('should generate a link to school-specific course path when schoolId is available', () => {
        render(<CourseCard course={basicCourse} />);
        const link = screen.getByRole('link');
        expect(link).toHaveAttribute('href', '/school/admin/school-123/courses/123');
    });

    it('should generate a link to org-specific course path when org_id is available', () => {
        render(<CourseCard course={courseWithOrgId} />);
        const link = screen.getByRole('link');
        expect(link).toHaveAttribute('href', '/school/admin/789/courses/456');
    });

    it('should generate a link to school path for learner courses', () => {
        render(<CourseCard course={courseWithRole} />);
        const link = screen.getByRole('link');
        expect(link).toHaveAttribute('href', '/school/test-org?course_id=789&cohort_id=123');
    });

    it('should generate a link to general course path when no school or org context is available', () => {
        // Override useParams to return no school ID
        require('next/navigation').useParams.mockReturnValue({});

        render(<CourseCard course={basicCourse} />);
        const link = screen.getByRole('link');
        expect(link).toHaveAttribute('href', '/courses/123');
    });

    it('should apply a border color based on course ID', () => {
        render(<CourseCard course={basicCourse} />);
        const card = screen.getByText('Test Course').closest('div');

        // Validate that a border class is applied
        // We can't test the exact class since it's dynamically generated,
        // but we can check that one of the border classes is present
        expect(card).toHaveClass('border-b-2');
        expect(card).toMatchSnapshot();

        // We can either validate against all possible border classes
        const possibleBorderClasses = [
            'border-purple-500',
            'border-green-500',
            'border-pink-500',
            'border-yellow-500',
            'border-blue-500',
            'border-red-500',
            'border-indigo-500',
            'border-orange-500'
        ];

        const hasBorderClass = possibleBorderClasses.some(className =>
            card?.classList.contains(className)
        );

        expect(hasBorderClass).toBe(true);
    });

    it('should handle string IDs for course correctly', () => {
        const courseWithStringId = {
            id: 'course-abc',
            title: 'Course with String ID'
        };

        render(<CourseCard course={courseWithStringId} />);
        const link = screen.getByRole('link');
        expect(link).toHaveAttribute('href', '/school/admin/school-123/courses/course-abc');
    });

    it('should show delete button when in admin view', () => {
        render(<CourseCard course={basicCourse} />);

        // Delete button should be present but initially not visible (opacity-0)
        const deleteButton = screen.getByLabelText('Delete course');
        expect(deleteButton).toBeInTheDocument();
        expect(deleteButton).toHaveClass('opacity-0');
    });

    it('should not show delete button when not in admin view', () => {
        // Override useParams to return no school ID
        require('next/navigation').useParams.mockReturnValue({});

        render(<CourseCard course={basicCourse} />);

        // Delete button should not be present
        const deleteButton = screen.queryByLabelText('Delete course');
        expect(deleteButton).not.toBeInTheDocument();
    });

    it('should open confirmation dialog when delete button is clicked', () => {
        render(<CourseCard course={basicCourse} />);

        // Click delete button
        fireEvent.click(screen.getByLabelText('Delete course'));

        // Confirmation dialog should be visible - using role to find specific elements
        const dialogHeading = screen.getByRole('heading', { name: 'Delete course' });
        expect(dialogHeading).toBeInTheDocument();
        expect(screen.getByText(/All the modules and tasks will be permanently deleted/)).toBeInTheDocument();
    });

    it('should close confirmation dialog when cancel is clicked', () => {
        render(<CourseCard course={basicCourse} />);

        // Open dialog
        fireEvent.click(screen.getByLabelText('Delete course'));
        const dialogHeading = screen.getByRole('heading', { name: 'Delete course' });
        expect(dialogHeading).toBeInTheDocument();

        // Click cancel
        fireEvent.click(screen.getByText('Cancel'));

        // Dialog should be closed
        expect(screen.queryByRole('heading', { name: 'Delete course' })).not.toBeInTheDocument();
    });

    it('should call API to delete course when confirmed', async () => {
        const onDeleteMock = jest.fn();
        render(<CourseCard course={basicCourse} onDelete={onDeleteMock} />);

        // Open dialog
        fireEvent.click(screen.getByLabelText('Delete course'));

        // Click delete button - use the correct button text "Delete" not "Delete course"
        const deleteButton = screen.getByRole('button', { name: 'Delete' });
        fireEvent.click(deleteButton);

        // Should show loading state - check for the spinner element within the button
        const buttonAfterClick = screen.getByRole('button', { name: 'Delete' });
        expect(buttonAfterClick.querySelector('.animate-spin')).toBeInTheDocument();

        // Should call fetch with correct URL and method
        expect(mockFetch).toHaveBeenCalledWith(
            'https://api.example.com/courses/123',
            expect.objectContaining({
                method: 'DELETE',
                headers: expect.objectContaining({
                    'Content-Type': 'application/json',
                })
            })
        );

        // Wait for operation to complete
        await waitFor(() => {
            // onDelete should have been called with course ID
            expect(onDeleteMock).toHaveBeenCalledWith(123);

            // Dialog should be closed
            expect(screen.queryByRole('heading', { name: 'Delete course' })).not.toBeInTheDocument();
        });
    });

    it('should show error message when delete fails', async () => {
        // Mock failed response
        mockFetch.mockResolvedValue({
            ok: false,
            status: 500,
            statusText: 'Internal Server Error'
        });

        render(<CourseCard course={basicCourse} />);

        // Open dialog
        fireEvent.click(screen.getByLabelText('Delete course'));

        // Click delete button - use a more specific selector
        const deleteButton = screen.getByRole('button', { name: 'Delete' });
        fireEvent.click(deleteButton);

        // Wait for error message
        await waitFor(() => {
            expect(screen.getByText(/An error occurred while deleting the course/)).toBeInTheDocument();
        });

        // Dialog should still be open
        expect(screen.getByRole('heading', { name: 'Delete course' })).toBeInTheDocument();
    });

    it('should prevent event propagation when delete button is clicked', () => {
        // Use a simpler approach to test event prevention
        render(<CourseCard course={basicCourse} />);

        // Get the delete button
        const deleteButton = screen.getByLabelText('Delete course');

        // We'll test event handling more implicitly
        // Just ensure clicking the button opens the dialog
        fireEvent.click(deleteButton);

        // Dialog should be open
        expect(screen.getByRole('heading', { name: 'Delete course' })).toBeInTheDocument();

        // And the link shouldn't have been navigated to
        // We know this because the test doesn't throw an error about navigation
    });
}); 