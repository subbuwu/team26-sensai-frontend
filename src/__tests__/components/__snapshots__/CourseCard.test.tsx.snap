// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCard Component should apply a border color based on course ID 1`] = `
<div
  class="bg-[#1A1A1A] text-gray-300 rounded-lg p-6 h-full transition-all hover:opacity-90 cursor-pointer border-b-2 border-yellow-500 border-opacity-70"
>
  <h2
    class="text-xl font-light mb-2"
  >
    Test Course
  </h2>
</div>
`;
